### DISCUSSION TITLE
zsh eval not work in zshrc with zoxide

### DISCUSSION THREAD

**Discussion by tshu-w** [marlonrichert/zsh-snap#93] [Q&A] [answered]
I have put the following lines in my zshrc, 
```
# zoxide
(( $+commands[zoxide] )) && znap eval zoxide "$(zoxide init --cmd j zsh)" && echo "yes"
```
however `zoxide` seems not to init correctly.
<img width="744" alt="Screen Shot 2021-07-31 at 16 13 41" src="https://user-images.githubusercontent.com/13161779/127733616-5d362cff-b8c1-4fff-a9a6-258976bac81f.png">

If I run in `znap eval` in command line and it works
```
$ znap eval zoxide "$(zoxide init --cmd j zsh)"
```
<img width="951" alt="Screen Shot 2021-07-31 at 16 20 27" src="https://user-images.githubusercontent.com/13161779/127733808-5942e219-14d1-4e6a-bc87-8a4d87b28b0a.png">

**Comment by marlonrichert** [comment, marked as answer]
> znap eval zoxide "$(zoxide init --cmd j zsh)"

Zsh evaluates expressions in `"`double quotes`"` _immediately_ upon declaration. You should use `'`single quotes`'` instead. 

Additionally, when using `znap eval`, you should leave out the `$(`…`)`.

**Reply by tshu-w** [reply]
@marlonrichert Thanks! I'm not very familiar with these, and I find I should use `znap eval zoxide "zoxide init --cmd j zsh"`. `znap eval zoxide '$(zoxide init --cmd j zsh)'` seems to be a bit problematic as well.

**Reply by marlonrichert** [reply]
> `znap eval zoxide '$(zoxide init --cmd j zsh)'` seems to be a bit problematic as well.

How?

**Reply by tshu-w** [reply]
I don't know. But I just put it in my zshrc and it's not work, but `znap eval zoxide "zoxide init --cmd j zsh"` works.

**Reply by marlonrichert** [reply]
> it's not work

In exactly what way does it "not work"? Please be more specific.

I just installed `zoxide`, add `znap eval zoxide 'zoxide init --cmd j zsh'` to my `.zshrc`, then restarted my shell and it works fine for me. I can now use `j` to jump to recent dirs.

**Reply by tshu-w** [reply]
@marlonrichert Yes `znap eval zoxide 'zoxide init --cmd j zsh'` works, have you tried `znap eval zoxide '$(zoxide init --cmd j zsh)'`, not work mean I cannot press j after start a new session.

**Reply by marlonrichert** [reply]
> have you tried `znap eval zoxide '$(zoxide init --cmd j zsh)'`, not work mean I cannot press j after start a new session.

When using `znap eval`, you should leave out `$(`…`)`. `znap eval zoxide 'zoxide init --cmd j zsh'` is the only correct way to do it. (I now updated my answer, above, to include this.)

**Reply by tshu-w** [reply]
@marlonrichert What about https://github.com/marlonrichert/zcolors#installation `znap eval zcolors "zcolors ${(q)LS_COLORS}"`? should we use `znap eval zcolors 'zcolors ${(q)LS_COLORS}'`

**Reply by marlonrichert** [reply]
No, you should use `znap eval zcolors "zcolors ${(q)LS_COLORS}"`, so that the `eval` cache will get invalidated whenever the value of `$LS_COLORS` changes. 🙂

Parameter expansion is cheap/fast. So, that should be in `"`double quotes`"`, so that the cache invalidates when the parameter value changes.

Getting output from a command, however, is slower/more expensive. That's why we want to cache that output, instead of regenerating it every time the shell starts up. Therefore, that should be in `'`single quotes`'`, so it's evaluated only once and then gets cached.

**Reply by tshu-w** [reply]
Thus make sense. Thx for your detailed explanation.

**Reply by marlonrichert** [reply]
You're welcome. 🙂

PS: If you enjoy using my software, please consider [sponsoring me](https://github.com/sponsors/marlonrichert).
